######
## Script name: filter_consensus.py
## Date: August 2023
## Author: Matteo Carrara (NEXUS Personalized Health Technologies)
##
## Description: Part of the Revseq pipeline. Loads the 
##        Filters the multi-FASTA containing consensus for all available viral sequences.
##        Possible filters are: only the top expressed (by RPKM proportions),
##        all top expressed (by RPKM proportions), or no filter
######

import pandas as pd
import argparse
from Bio import SeqIO


# Script
if __name__ == '__main__':
	# Parse input args
    parser = argparse.ArgumentParser(description='Filters the full consensus list to keep only the consensus sequences of interest', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('--ref_table', required=True, type=str, help='the bed file listing all the reference viral genomes')
    parser.add_argument('--assignment', required=True, type=str, help='Substrain assignment table generated by assign_virus.py')
    parser.add_argument('--consensus', required=True, type=str, help='Consensus FASTA file generated by BCFTools')
    parser.add_argument('--output', required=True, type=str, help='File where to write the filtered consensus')
    parser.add_argument('--consensus_type', required=True, type=str, help='Filtering type for consensus. Accepted values are "all", "outliers" and "top"')
    parser.add_argument('--readcount', required=True, type=str, help='Read number from the remove_duplicates step')
    args = parser.parse_args()

    refs = pd.read_table(args.ref_table, header=None)
    refs = refs.rename(columns={0: "id", 1: "ref_start", 2: "ref_end", 3: "substrain_name"})
    with open(args.readcount) as f:
        readcount=int(f.read().strip())
    assignment = pd.read_table(args.assignment, header=0, sep="\t")
    assignment=assignment.rename(columns={"Unnamed: 0": "name"})
    if args.consensus_type == "top":
        print("Calculating consensus only for the top strain")
        filter_values = assignment.loc[assignment["rpkm_proportions"] == max(assignment["rpkm_proportions"])]["name"].to_list()
    elif args.consensus_type == "outliers":
        print("Calculating consensus for all outlier strains")
        filter_values = assignment.loc[assignment["outlier"] == "*"]["name"].to_list()
    elif args.consensus_type == "all":
        print("Calculating consensus all strains")
        filter_values = assignment["name"].to_list()
    else:
        sys.exit("Error: cannot recognize the value assigned to 'consensus_type': " + args.consensus_type)

    if len(filter_values) == 0:
        print("WARNING: no reads available after dehumanization and filtering, resulting in no assignment.")
        with open(args.output, "a") as out_file:
            out_file.write("")
    else:
        assignment = assignment.loc[assignment["name"].isin(filter_values)]
        regions = refs.loc[refs["substrain_name"].isin(filter_values)]["id"].to_list()

        consensus = SeqIO.parse(open(args.consensus),'fasta')
        fasta_sequences = [record for record in consensus if record.id in regions ]
        with open(args.output, "a") as out_file:
            print("writing the consensus")
            for seq in fasta_sequences:
                SeqIO.write(seq, out_file, "fasta")