from snakemake.utils import min_version
from snakemake.utils import validate

# do not allow subdirectories as part of `sample` or `i` wildcards
wildcard_constraints:
    sample="[^/]+",

# Define minimum Snakemake version
min_version("6.12.1")

# Include Config file
configfile: "config/config.yaml"


# Include rules
include: "rules/misc_snake.smk"
include: "rules/preprocessing.smk"
include: "rules/dehuman.smk"
include: "rules/old_dehuman.smk"
include: "rules/revseq_basic.smk"
include: "rules/qc.smk"

# include local rules
localrules:
    all,

# final rule of pipeline
# defines output
rule all:
    input:
        expand(config["inputOutput"]["output_dir"]+"/{sample}/merge_lanes/{sample}_merged_R1.fastq.gz", sample = sample_ids, lane = lane_ids),
        expand(config["inputOutput"]["output_dir"]+"/{sample}/merge_lanes/{sample}_merged_R2.fastq.gz", sample = sample_ids, lane = lane_ids),
        #expand(config["inputOutput"]["output_dir"]+"/{sample}/bwa/mapped_reads.bam", sample = sample_ids),
        #expand(config["inputOutput"]["output_dir"]+"/{sample}/sort/sorted_reads.bam", sample = sample_ids),
        expand(config["inputOutput"]["output_dir"]+"/{sample}/pileup/pileup.txt", sample = sample_ids),
        #expand(config["inputOutput"]["output_dir"]+"/{sample}/pair_freqs/pair_counts.pkl.gz", sample = sample_ids),
        #expand(config["inputOutput"]["output_dir"]+"/{sample}/consensus/consensus.fasta", sample = sample_ids),
        expand(config["inputOutput"]["output_dir"]+"/{sample}/dehuman/dehuman.cram", sample = sample_ids),
        #expand(config["inputOutput"]["output_dir"]+"/{sample}/filter_host_reads/{sample}_dehuman.cram", sample = sample_ids),
        expand(config["inputOutput"]["output_dir"]+"/{sample}/fastqc/{sample}_L001_R1_001_fastqc.zip", sample = sample_ids),#, lane = lane_ids),
        expand(config["inputOutput"]["output_dir"]+"/{sample}/samtoolsstats/{sample}_samtools_stats.txt", sample=sample_ids),
        expand(config["inputOutput"]["output_dir"]+"/{sample}/rseqc/{sample}_bam_stat.txt", sample=sample_ids),
        config["inputOutput"]["output_dir"]+"/multiqc/multiqc_report.html",
    output:
        config["inputOutput"]["output_dir"]+"/complete.txt",
    shell:
        "date > {output}"




